futures-util/src/io/read_line.rs

```diff
@@ -29,20 +29,31 @@ impl<'a, R: AsyncBufRead + ?Sized + Unpin> ReadLine<'a, R> {
     }
 }
 
+pub(super) fn read_line_internal<R: AsyncBufRead + ?Sized>(
+    reader: Pin<&mut R>,
+    buf: &mut String,
+    bytes: &mut Vec<u8>,
+    read: &mut usize,
+    cx: &mut Context<'_>,
+) -> Poll<io::Result<usize>> {
+    let ret = ready!(read_until_internal(reader, b'\n', bytes, read, cx));
+    if str::from_utf8(&bytes).is_err() {
+        Poll::Ready(ret.and_then(|_| {
+            Err(io::Error::new(io::ErrorKind::InvalidData, "stream did not contain valid UTF-8"))
+        }))
+    } else {
+        debug_assert!(buf.is_empty());
+        debug_assert_eq!(*read, 0);
+        unsafe { mem::swap(buf.as_mut_vec(), bytes); }
```

Comment:
nit: would you mind moving the `unsafe { ... }` just around the `as_mut_vec()` (inside the `mem::swap`) and leaving a comment explaining that this is safe because we've already validated that `bytes` is valid UTF8 above?